commp.go

```diff
@@ -280,60 +287,71 @@ func (cp *Calc) addLayer(myIdx uint) {
 	cp.layerQueues[myIdx+1] = make(chan []byte, layerQueueDepth)
 
 	go func() {
-		var chunkHold []byte
+		var twinHold []byte
 
 		for {
-
-			chunk, queueIsOpen := <-cp.layerQueues[myIdx]
+			slab, queueIsOpen := <-cp.layerQueues[myIdx]
 
 			// the dream is collapsing
 			if !queueIsOpen {
+				defer func() { twinHold = nil }()
 
 				// I am last
 				if myIdx == MaxLayers || cp.layerQueues[myIdx+2] == nil {
-					cp.resultCommP <- chunkHold
+					cp.resultCommP <- append(make([]byte, 0, 32), twinHold[0:32]...)
 					return
 				}
 
-				if chunkHold != nil {
-					cp.hash254Into(
-						cp.layerQueues[myIdx+1],
-						chunkHold,
-						stackedNulPadding[myIdx],
-					)
+				if twinHold != nil {
+					copy(twinHold[32:64], stackedNulPadding[myIdx])
+					cp.hashSlab254(0, twinHold[0:64])
+					cp.layerQueues[myIdx+1] <- twinHold[0:64:64]
 				}
 
 				// signal the next in line that they are done too
 				close(cp.layerQueues[myIdx+1])
 				return
 			}
 
-			if chunkHold == nil {
-				chunkHold = chunk
-			} else {
-
-				// We are last right now
-				// n.b. we will not blow out of the preallocated layerQueues array,
-				// as we disallow Write()s above a certain threshold
-				if cp.layerQueues[myIdx+2] == nil {
-					cp.addLayer(myIdx + 1)
-				}
+			var pushedWork bool
+
+			switch {
+			case len(slab) > 1<<(5+myIdx):
+				cp.hashSlab254(myIdx, slab)
+				cp.layerQueues[myIdx+1] <- slab
+				pushedWork = true
+			case twinHold != nil:
+				copy(twinHold[32:64], slab[0:32])
+				cp.hashSlab254(0, twinHold[0:64])
+				cp.layerQueues[myIdx+1] <- twinHold[0:32:64]
+				pushedWork = true
+				twinHold = nil
+			default:
+				twinHold = slab[0:32:64]
+			}
 
-				cp.hash254Into(cp.layerQueues[myIdx+1], chunkHold, chunk)
-				chunkHold = nil
+			// Check whether we need another worker
+			//
+			// n.b. we will not blow out of the preallocated layerQueues array,
+			// as we disallow Write()s above a certain threshold
+			if pushedWork && cp.layerQueues[myIdx+2] == nil {
+				cp.addLayer(myIdx + 1)
 			}
 		}
 	}()
 }
 
-func (cp *Calc) hash254Into(out chan<- []byte, half1ToOverwrite, half2 []byte) {
+func (cp *Calc) hashSlab254(layerIdx uint, slab []byte) {
 	h := shaPool.Get().(hash.Hash)
-	h.Reset()
-	h.Write(half1ToOverwrite)
-	h.Write(half2)
-	d := h.Sum(half1ToOverwrite[:0]) // callers expect we will reuse-reduce-recycle
-	d[31] &= 0x3F
-	out <- d
+
+	stride := 1 << (5 + layerIdx)
```

Comment:
sorry, i am not quite following the index + 5 choice here -- mind walking me through it?

Reply:
The leaves are 32 Bytes ( 1 << 5 )
On layer0 we hash every 2 neighbouring leaves - pos0, pos1, pos2, pos3, .... and the results are written in pos0, ..., pos2, ... 
On the next layer pos0 and pos2 are hashed and the result is stored in pos0 -> stride 64, etc